src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -812,101 +731,204 @@ public int getCurrentRunningCount() {
-         * available workers BEFORE stopping.
-         */
-        w.stopIfRunning();
-        currentPoolSize--;
-        // it may not always be here, but it sometimes can (for example when a worker is interrupted)
-        availableWorkers.remove(w);
+    protected void addWorkerToIdleChain(Worker worker) {
+      while (true) {
+        Worker casWorker = idleWorker.get();
+        // we can freely set this value until we get into the idle linked queue
+        worker.nextIdleWorker = casWorker;
+        if (idleWorker.compareAndSet(casWorker, worker)) {
+          break;
+        }
       }
     }
-    
+
     /**
-     * Called by the worker after it completes a task.  This is so that we can run any after task 
-     * cleanup, and make sure that the worker is now available for future tasks.
+     * Invoked when a worker becomes idle.  This will provide another task for that worker, or 
+     * block until a task is either ready, or the worker should be shutdown (either because pool 
+     * was shut down, or max pool size changed).
      * 
-     * @param worker worker that is now idle and ready for more tasks
+     * @param worker Worker which is now idle and ready for a task
+     * @return Task that is ready for immediate execution
      */
-    protected void workerDone(Worker worker) {
-      synchronized (workersLock) {
-        if (shutdownFinishing || currentPoolSize > maxPoolSize) {
-          killWorker(worker);
-        } else {
-          // always add to the front so older workers are at the back
-          availableWorkers.addFirst(worker);
-          
-          if (waitingForWorker) {
-            workersLock.notify();
+    public TaskWrapper workerIdle(Worker worker) {
+      boolean interruptedChecked = false;
+      boolean queued = false;
+      try {
+        idle: while (true) {
+          int casPoolSize;
+          if (shutdownFinishing) {
+            currentPoolSize.decrementAndGet();
+            worker.stopIfRunning();
+            return null;
+          } else if ((casPoolSize = currentPoolSize.get()) > maxPoolSize) {
+            if (currentPoolSize.compareAndSet(casPoolSize, casPoolSize - 1)) {
+              worker.stopIfRunning();
+              return null;
+            }
+          } else {
+            // ready to poll task, we loop here to avoid pool state recheck logic unless we have to block
+            while (true) {
+              TaskWrapper nextTask = queueManager.getNextTask();
+              if (nextTask == null) {
+                if (queued) {
+                  // we can only park after we have queued, then checked again for a result
+                  LockSupport.park();
+                  continue idle;
+                } else {
+                  addWorkerToIdleChain(worker);
+                  queued = true;
+                }
+              } else {
+                /* TODO - right now this has a a deficiency where a recurring period task can cut in 
```

Comment:
I see what you mean.  A possible fix might be getting/checking a run counter so we can tell if it has had a run while we where checking to see if we can run it.  If the counter is not what we got originally we just skip it.  Most normal tasks would just go from 0 to 1.  Recurring would just keep going up every time they run.